qiskit_experiments/curve_analysis/curve_analysis.py

```diff
@@ -380,53 +379,75 @@ def _objective(_params):
 
-        # Run data processing
-        processed_data = self._run_data_processing(
-            raw_data=experiment_data.data(),
-            models=self._models,
-        )
+        Args:
+            curve_data: Scatter data table containing all data points.
 
-        if self.options.plot and self.options.plot_raw_data:
-            for model in self._models:
-                sub_data = processed_data.get_subset_of(model._name)
+        Returns:
+            A list of figures.
+        """
+        for model_name, data in list(curve_data.groupby("model_name")):
+            # Plot raw data scatters
+            if self.options.plot_raw_data:
+                raw_data = data[data.format == "raw"]
                 self.plotter.set_series_data(
-                    model._name,
-                    x=sub_data.x,
-                    y=sub_data.y,
+                    series_name=model_name,
+                    x=raw_data.xval.to_numpy(),
+                    y=raw_data.yval.to_numpy(),
                 )
-
-        # Format data
-        formatted_data = self._format_data(processed_data)
-        if self.options.plot:
-            for model in self._models:
-                sub_data = formatted_data.get_subset_of(model._name)
+            # Plot formatted data scatters
+            formatted_data = data[data.format == "fit-ready"]
+            self.plotter.set_series_data(
+                series_name=model_name,
+                x_formatted=formatted_data.xval.to_numpy(),
+                y_formatted=formatted_data.yval.to_numpy(),
+                y_formatted_err=formatted_data.yerr.to_numpy(),
+            )
+            # Plot fit lines
+            line_data = data[data.format == "fit"]
+            if len(line_data) == 0:
+                continue
+            self.plotter.set_series_data(
+                series_name=model_name,
+                x_interp=line_data.xval.to_numpy(),
+                y_interp=line_data.yval.to_numpy(),
+            )
+            fit_stdev = line_data.yerr.to_numpy()
+            if np.isfinite(fit_stdev).all():
                 self.plotter.set_series_data(
-                    model._name,
-                    x_formatted=sub_data.x,
-                    y_formatted=sub_data.y,
-                    y_formatted_err=sub_data.y_err,
+                    series_name=model_name,
+                    y_interp_err=fit_stdev,
                 )
 
-        # Run fitting
-        fit_data = self._run_curve_fit(
-            curve_data=formatted_data,
-            models=self._models,
-        )
+        return [self.plotter.figure()]
+
+    def _run_analysis(
+        self,
+        experiment_data: ExperimentData,
+    ) -> Tuple[List[AnalysisResultData], List["pyplot.Figure"]]:
+        analysis_results = []
+        figures = []
+
+        # Prepare for fitting
+        self._initialize(experiment_data)
+
+        curve_data = self._format_data(self._run_data_processing(experiment_data.data()))
```

Comment:
This curve_data object is now `ScatterTable`. This table can store both processed_data and formatted_data and code becomes much simpler.